src/interop/js-interop/js-types.md

```diff
@@ -0,0 +1,212 @@
+limitation. Look at the specific conversion function for more details.
+
+## Requirements on `external` declarations and `Function.toJS`
+
+In order to ensure type safety and consistency, the compiler places requirements
+on what types can flow into and out of JS. Passing arbitrary Dart values into JS
+is not allowed. Instead, the compiler requires users to use a compatible interop
+type like a JS type or a primitive, which would then be implicitly converted by
+the compiler. For example, these would be allowed:
+
+{:.good}
+```dart
+@JS()
+external void primitives(String a, int b, double c, num d, bool e);
+```
+
+{:.good}
+```dart
+@JS()
+external JSArray jsTypes(JSObject _, JSString __);
+```
+
+{:.good}
+```dart
+extension type InteropType(JSObject _) {}
+
+@JS()
+external InteropType get interopType;
+```
+
+whereas these would return an error:
+
+{:.bad}
+```dart
+@JS()
+external Function get function;
+```
+
+{:.bad}
+```dart
+@JS()
+external set list(List _);
+```
+
+These same requirements exist when you use [`Function.toJS`] to make a Dart
+function callable in JS. The values that flow into and out of this callback must
+be a compatible interop type or a primitive.
+
+If you use a Dart primitive e.g. `String`, an implicit conversion happens in the
+compiler to convert that value from a JS value to a Dart value. If performance
+is critical and you don’t need to examine the contents of the string, then using
+`JSString` instead to avoid the conversion cost may make sense.
+
+## Compatibility, type checks, and casts
+
+As mentioned above, the representation type of JS types may differ based on the
+compiler. This affects runtime type-checking and casts. Therefore, almost always
+avoid `is` checks where the value is an interop type or where the target type is
+an interop type. Avoid casts between Dart types and interop types. In order to
+type-check a JS value, use an interop member like [`typeofEquals`] or
+[`instanceOfString`] that examines the JS value itself.
+
+Bad:
+
+{:.bad}
+```dart
+void f(JSAny a) {
+  if (a is String) { … }
+}
+```
+
+{:.bad}
+```dart
+void f(JSObject o) {
+  if (o is JSObject) { … }
+}
+```
+
+{:.bad}
+```dart
+void f(JSString s) {
+  s as String;
+}
+```
+
+Okay:
+
+{:.good}
+```dart
+void f(JSAny a) {
+  // Here `a` is verified to be a JS function, so the cast is okay.
+  if (a.typeofEquals('function')) {
+    a as JSFunction;
+  }
+}
+```
+
+Dart may add lints to make runtime checks with JS interop types easier to avoid.
+
+## `null` vs `undefined`
+
+JS has both a `null` and an `undefined` value. This is in contrast with Dart,
+which only has `null`. In order to make JS values more ergonomic to use, if an
+interop member were to return either JS `null` or `undefined`, the compiler maps
+these values to Dart `null`. Therefore a member like:
+
+```dart
+@JS()
+external JSObject? get value;
+```
+
+can be interpreted as returning a JS object, JS `null`, or `undefined`.
```

Comment:
The example in the middle of the sentence is kind of weird, so no big deal but if it makes sense try:

````suggestion
these values to Dart `null`. Therefore a member like `value` in the following example
can be interpreted as returning a JS object, JS `null`, or `undefined`:

```dart
@JS()
external JSObject? get value;
```
````

Is `value` even the member you're referring to? 😅  Whatever it is, hopefully you get the idea!

Reply:
It is, so this is clearer, thanks!